src/shogun/base/SGObject.cpp

```diff
@@ -56,59 +57,120 @@
 namespace shogun
 {
 
-	typedef std::map<BaseTag, AnyParameter> ParametersMap;
+	typedef std::vector<std::pair<BaseTag, AnyParameter>> ParametersMap;
 	typedef std::unordered_map<std::string_view, std::string_view> ObsParamsList;
 
+#define SORTED 0
```

Comment:
I guess this bunch of `#if` and `#else` will be removed once we choose the final implementation.

Looking at the benchmarks, I guess we could go with the unsorted option (unless we prefer to have the
parameter sorted for some reason). Since I don't think we will ever have like 1000 parameters registered we don't need to sort.

Reply:
yeah actually no... i've refactored a bit the whole SGObject::Self. this way it's easier to add a new type of backend. like a distributed key-value store? ;)
i opted to use template based solution instead of vtable, hoping that it's less amount of overhead.